hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/OzoneClientUtils.java

```diff
@@ -280,4 +290,53 @@ public static void deleteSnapshot(ObjectStore objectStore,
           exception.getMessage());
     }
   }
+
+  public static void setReplication(OzoneConfiguration conf, OzoneBucket bucket,
+      String keyName, short replication) throws IOException {
+    OzoneKeyDetails keyDetails;
+    try {
+      keyDetails = bucket.getKey(keyName);
+    } catch (OMException ome) {
+      // if key does not exist, do nothing
+      if (ome.getResult() == KEY_NOT_FOUND) {
+        return;
+      }
+      throw ome;
+    }
+    ReplicationConfig newReplication =
+        resolveClientSideReplicationConfig(replication, null,
+            null, conf);
+    if (newReplication == null) {
+      throw new IOException(
+          "Replication factor of " + replication + " not supported");
+    }
+    if (newReplication.getReplication()
+        .equals(keyDetails.getReplicationConfig().getReplication())) {
+      return;
+    }
+    try (InputStream inputStream = readFile(bucket, keyName);
+        OzoneOutputStream outputStream = bucket.rewriteKey(keyName,
+            keyDetails.getDataSize(), keyDetails.getGeneration(),
+            newReplication, keyDetails.getMetadata())) {
+      int chunkSize = (int) conf.getStorageSize(OZONE_SCM_CHUNK_SIZE_KEY,
+          OZONE_SCM_CHUNK_SIZE_DEFAULT, StorageUnit.BYTES);
+      IOUtils.copyBytes(inputStream, outputStream, chunkSize);
+    }
```

Comment:
I agree with @jojochuang here.

Also what happens when a user Ctrl-C during a command like this on a large key? Could the data be lost?

Reply:
Why would data be lost?  `rewriteKey` is like overwriting a key normally using `ozone sh key put`, except that it commits the change only if the key was not updated by others in the meantime.

Reply:
I think what @smengcl means is that is in between the execution of 
`IOUtils.copyBytes(inputStream, outputStream, chunkSize)` if the process is killed  with say a SIGTERM, what  would happen? 
Since this is defined in a try with resource block the outputstream will be closed on exception committing the partially written key to OM. 

But don't we have this problem with a regular write too? If the client abruptly shuts down then it could lead to same situation. Are you saying that such behaviour is not expected from setrep as the expectation is not to rewrite ( same as @jojochuang's comment)

Reply:
> Why would data be lost? `rewriteKey` is like overwriting a key normally using `ozone sh key put`, except that it commits the change only if the key was not updated by others in the meantime.

@adoroszlai Ah I see. Since `rewriteKey()` is atomic, is it true that if the rewrite is interrupted (not properly closed/committed, or updated by another client), the new replication factor won't take effect at all?

Reply:
If `KeyOutputStream` is not closed, the rewrite will not be completed:

```
[ERROR] org.apache.hadoop.fs.ozone.TestOFS.testSetReplication  Time elapsed: 0.592 s  <<< FAILURE!
org.opentest4j.AssertionFailedError: expected: <THREE> but was: <ONE>
```

@sadanand48 it may be useful to add a negative test like that.